Assertion `!isa<ProtocolDecl>(this) && "Self-conformances are only found by the higher-level " "ModuleDecl::lookupConformance() entry point"' failed.

**Describe the bug**
Reduced test code:

```swift
import Foundation

struct SQLiteKeyedDecodingContainer<Key>: KeyedDecodingContainerProtocol where Key: CodingKey {
   func superDecoder() throws -> Decoder { throw DecodingError.keyNotFound(CodingKey("super"), DecodingError.Context(codingPath: codingPath, debugDescription: "There is no super container")) }
}
```

results in

```
swift-frontend: /home/me/llvm/swift-source/swift/lib/AST/ProtocolConformance.cpp:1106: bool swift::NominalTypeDecl::lookupConformance(swift::ProtocolDecl *, SmallVectorImpl<swift::ProtocolConformance *> &) const: Assertion `!isa<ProtocolDecl>(this) && "Self-conformances are only found by the higher-level " "ModuleDecl::lookupConformance() entry point"' failed.
Stack dump:
0.      Program arguments: /home/me/llvm/opt/usr/bin/swift-frontend -frontend -interpret SQLiteDecoder.swift -disable-objc-interop -color-diagnostics -new-driver-path /home/me/llvm/opt/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /home/me/llvm/opt/usr/lib/swift -module-name SQLiteDecoder
1.      Swift version 5.8-dev (LLVM c205842748e1789, Swift 1f8a5d6045262bb)
2.      Compiling with the current language version
3.      While evaluating request TypeCheckSourceFileRequest(source_file "SQLiteDecoder.swift")
4.      While evaluating request TypeCheckFunctionBodyRequest(SQLiteDecoder.(file).SQLiteKeyedDecodingContainer.superDecoder()@SQLiteDecoder.swift:4:9)
5.      While type-checking statement at [SQLiteDecoder.swift:4:42 - line:4:192] RangeText="{ throw DecodingError.keyNotFound(CodingKey("super"), DecodingError.Context(codingPath: codingPath, debugDescription: "There is no super container")) "
6.      While type-checking statement at [SQLiteDecoder.swift:4:44 - line:4:190] RangeText="throw DecodingError.keyNotFound(CodingKey("super"), DecodingError.Context(codingPath: codingPath, debugDescription: "There is no super container")"
7.      While type-checking expression at [SQLiteDecoder.swift:4:50 - line:4:190] RangeText="DecodingError.keyNotFound(CodingKey("super"), DecodingError.Context(codingPath: codingPath, debugDescription: "There is no super container")"
8.      While type-checking-target starting at SQLiteDecoder.swift:4:64
 #0 0x000000000657f0e3 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/me/llvm/opt/usr/bin/swift-frontend+0x657f0e3)
 #1 0x000000000657cece llvm::sys::RunSignalHandlers() (/home/me/llvm/opt/usr/bin/swift-frontend+0x657cece)
 #2 0x000000000657f46f SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f31fbb69140 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13140)
 #4 0x00007f31fafe1ce1 raise ./signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f31fafcb537 abort ./stdlib/abort.c:81:7
 #6 0x00007f31fafcb40f get_sysdep_segment_value ./intl/loadmsgcat.c:509:8
 #7 0x00007f31fafcb40f _nl_load_domain ./intl/loadmsgcat.c:970:34
 #8 0x00007f31fafda662 (/lib/x86_64-linux-gnu/libc.so.6+0x31662)
 #9 0x0000000001da5ae2 (/home/me/llvm/opt/usr/bin/swift-frontend+0x1da5ae2)
#10 0x00000000014f45b6 (anonymous namespace)::PreCheckExpression::simplifyTypeConstructionWithLiteralArg(swift::Expr*) PreCheckExpr.cpp:0:0
#11 0x00000000014f20b2 (anonymous namespace)::PreCheckExpression::walkToExprPost(swift::Expr*) PreCheckExpr.cpp:0:0
#12 0x0000000001b9c0d8 (anonymous namespace)::Traversal::doIt(swift::Expr*) ASTWalker.cpp:0:0
#13 0x0000000001b9e999 (anonymous namespace)::Traversal::visitApplyExpr(swift::ApplyExpr*) ASTWalker.cpp:0:0
#14 0x0000000001b9be02 (anonymous namespace)::Traversal::doIt(swift::Expr*) ASTWalker.cpp:0:0
#15 0x0000000001b9bd03 swift::Expr::walk(swift::ASTWalker&) (/home/me/llvm/opt/usr/bin/swift-frontend+0x1b9bd03)
#16 0x00000000014f0516 swift::constraints::ConstraintSystem::preCheckExpression(swift::Expr*&, swift::DeclContext*, bool, bool) (/home/me/llvm/opt/usr/bin/swift-frontend+0x14f0516)
#17 0x00000000014f03a9 swift::constraints::ConstraintSystem::preCheckTarget(swift::constraints::SolutionApplicationTarget&, bool, bool) (/home/me/llvm/opt/usr/bin/swift-frontend+0x14f03a9)
#18 0x0000000001563283 swift::TypeChecker::typeCheckTarget(swift::constraints::SolutionApplicationTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/home/me/llvm/opt/usr/bin/swift-frontend+0x1563283)
#19 0x000000000156319d swift::TypeChecker::typeCheckExpression(swift::constraints::SolutionApplicationTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/home/me/llvm/opt/usr/bin/swift-frontend+0x156319d)
#20 0x000000000156308d swift::TypeChecker::typeCheckExpression(swift::Expr*&, swift::DeclContext*, swift::constraints::ContextualTypeInfo, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) (/home/me/llvm/opt/usr/bin/swift-frontend+0x156308d)
#21 0x00000000016346d1 swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) TypeCheckStmt.cpp:0:0
#22 0x0000000001633efc bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::Stmt>(swift::Stmt*&) TypeCheckStmt.cpp:0:0
#23 0x000000000163121e (anonymous namespace)::StmtChecker::typeCheckASTNode(swift::ASTNode&) TypeCheckStmt.cpp:0:0
#24 0x0000000001633fac swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) TypeCheckStmt.cpp:0:0
#25 0x0000000001632cfc bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) TypeCheckStmt.cpp:0:0
#26 0x00000000016320e1 swift::TypeCheckFunctionBodyRequest::evaluate(swift::Evaluator&, swift::AbstractFunctionDecl*) const (/home/me/llvm/opt/usr/bin/swift-frontend+0x16320e1)
#27 0x0000000001c9bc24 llvm::Expected<swift::TypeCheckFunctionBodyRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckFunctionBodyRequest>(swift::TypeCheckFunctionBodyRequest const&) crtstuff.c:0:0
#28 0x0000000001c9b959 llvm::Expected<swift::TypeCheckFunctionBodyRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckFunctionBodyRequest, (void*)0>(swift::TypeCheckFunctionBodyRequest const&) crtstuff.c:0:0
#29 0x0000000001c00d78 swift::TypeCheckFunctionBodyRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckFunctionBodyRequest>(swift::Evaluator&, swift::TypeCheckFunctionBodyRequest, swift::TypeCheckFunctionBodyRequest::OutputType) (/home/me/llvm/opt/usr/bin/swift-frontend+0x1c00d78)
#30 0x0000000001c00d2f swift::AbstractFunctionDecl::getTypecheckedBody() const (/home/me/llvm/opt/usr/bin/swift-frontend+0x1c00d2f)
#31 0x000000000166e136 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const (/home/me/llvm/opt/usr/bin/swift-frontend+0x166e136)
#32 0x00000000016704c8 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) crtstuff.c:0:0
#33 0x00000000016701e2 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) crtstuff.c:0:0
#34 0x000000000166df04 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) (/home/me/llvm/opt/usr/bin/swift-frontend+0x166df04)
#35 0x000000000166dee4 swift::performTypeChecking(swift::SourceFile&) (/home/me/llvm/opt/usr/bin/swift-frontend+0x166dee4)
#36 0x00000000006d9cd9 bool llvm::function_ref<bool (swift::SourceFile&)>::callback_fn<swift::CompilerInstance::performSema()::$_7>(long, swift::SourceFile&) Frontend.cpp:0:0
#37 0x00000000006ce33e swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<bool (swift::SourceFile&)>) (/home/me/llvm/opt/usr/bin/swift-frontend+0x6ce33e)
#38 0x00000000006ce287 swift::CompilerInstance::performSema() (/home/me/llvm/opt/usr/bin/swift-frontend+0x6ce287)
#39 0x000000000068c51e withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) FrontendTool.cpp:0:0
#40 0x000000000067d467 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) FrontendTool.cpp:0:0
#41 0x000000000067c328 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/home/me/llvm/opt/usr/bin/swift-frontend+0x67c328)
#42 0x00000000004c0935 swift::mainEntry(int, char const**) (/home/me/llvm/opt/usr/bin/swift-frontend+0x4c0935)
#43 0x00007f31fafccd0a __libc_start_main ./csu/../csu/libc-start.c:308:16
#44 0x00000000004bfaca _start (/home/me/llvm/opt/usr/bin/swift-frontend+0x4bfaca)
```


**Steps To Reproduce**
Steps to reproduce the behavior:
1. Create a dummy.swift that contains the reduced example code
2. swift dummy.swift

**Expected behavior**
Some kind of error message (the code is incorrect).

**Screenshots**
-/-

**Environment (please fill out the following information)**
 - OS: Debian Linux
 - Self compiled swiftc, LLVM c205842748e1789, Swift 1f8a5d6045262bb

**Additional context**
-/-

We correctly reject a call to `.init`, so it should be easy to figure out what's going wrong with `P(false)`:

```swift
foo.swift:7:3: error: type 'any P' cannot be instantiated
P.init(false)
~ ^
```

--- a/lib/Sema/PreCheckExpr.cpp
+++ b/lib/Sema/PreCheckExpr.cpp
@@ -2119,8 +2119,7 @@ Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
       return nullptr;
   }
 
-  SmallVector<ProtocolConformance *, 2> conformances;
-  return castTy->getAnyNominal()->lookupConformance(protocol, conformances)
+  return DC->getParentModule()->lookupConformance(castTy, protocol)
              ? CoerceExpr::forLiteralInit(getASTContext(), literal,
                                           call->getSourceRange(),
                                           typeExpr->getTypeRepr())

--- a/test/Constraints/protocols.swift
+++ b/test/Constraints/protocols.swift
@@ -533,3 +533,16 @@ func f1_58231(x: Int) -> P_58231 {
 func f2_58231(x: Int) -> P_58231? {
   return S_58231() // expected-error{{return expression of type 'S_58231' does not conform to 'P_58231'}}
 }
+
+// https://github.com/apple/swift/issues/61517
+
+protocol P_61517 {
+  init()
+  init(_: Bool)
+}
+
+_ = P_61517() // expected-error{{type 'any P_61517' cannot be instantiated}}
+_ = P_61517(false) // expected-error{{type 'any P_61517' cannot be instantiated}}
+
+_ = P_61517.init() // expected-error{{type 'any P_61517' cannot be instantiated}}
+_ = P_61517.init(false) // expected-error{{type 'any P_61517' cannot be instantiated}}

